### PR TITLE
Support for Overlay Controllers / EVPN Route servers & Flexible EVPN Overlay Peerings

### LINKED ISSUE #269: Add External Route Server / Route Reflector support in L3LS data model
**Issue Type**

- Role enhancement

**Is your feature request related to a problem? Please describe.**

For large deployments, it is a common design to use a set of devices dedicated to act as RR or RS and only reflect all the control plane information with no transit function.

**Describe the solution you'd like**

Add a specific role in `eos_l3ls_evpn` role to configure such design.

**Describe alternatives you've considered**

Do an additional play to leverage `eos_cli_config_gen` to model such a device. But it is not scalable and does not sort out connectivity management on the fabric side.

**Additional context**

Customer requirement

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

* Add new device type `overlay_controller`
* Add `evpn_overlay_controller_groups` to specify upstream EVPN overlay peers

TODO: Create / Adjust generated documentation for these new features.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #269
## Component(s) name
eos_l3ls_evpn

## Role Enchancements for Super Spine Support

The enchancement listed below are required to support bigger deployments with super-spines (5 stage CLOS).
5 stage CLOS fabric can be represented as multiple leaf-spine structures (called PODs - Point of Delivery) interconnected by super-spines.
The logic to deploy every leaf-spine POD fabric remains unchanged. The enchancement only adds logic required to provision spine-to-super-spine fabric.
Super-spines can be deployed as a single plane (typically chassis switches) or multiple planes.
Current AVD release supports single plane deployment only.

Only BGP underlay is supported for super-spine deployment. Spines in every POD must have unique AS per POD.

### Inventory Structure

The inventory must have a dedicated group for super-spines and every leaf-spine POD. Example:

```yaml
all:
  children:
    < DC-group-name >:
      children:
        < Super Spines group name >:
          hosts:
            < super-spine name >:
              ansible_host: < management IP >
            < super-spine name >:
              ansible_host: < management IP >
            ...
        < DC POD 1 group name >:
          children:
            < spines group >:
              <-- omitted -->
            < leaf group >:
              <-- omitted -->
```

### Additional Variables Required For Super Spine Deployment

Defaults:

```yaml
max_spine_to_super_spine_links: 1  # number of parallel links between spines and super-spines
```

Assigned to the DC group:

```yaml
dc_name: DC1  # data center fabric name
              # this variable is required to identify devices in the same DC in case of multi-DC setup

max_super_spines: 4  # maximum number of super-spines, changing this parameter affects address allocation

super_spine:
  platform: vEOS-LAB  # super-spine platform
  bgp_as: <super-spine BGP AS>
  nodes:
    SU-01:  # super-spine name
      id: 1
      mgmt_ip: 192.168.0.1/24
    <-- etc. -->

# IP address range for loopbacks for all super-spines in the DC,
# assigned as /32s
# Assign range larger then total super-spines
super_spine_loopback_network_summary: 192.168.100.0/24

# additional lines for super-spine BGP config
super_spine_bgp_defaults:
  #  - update wait-for-convergence
  #  - update wait-install
  - no bgp default ipv4-unicast
  - distance bgp 20 200 200
  - graceful-restart restart-time 300
  - graceful-restart
```

Assigned to Super Spine Group:

```yaml
type: super-spine  # identifies every host in the group as super-spine
```

Assigned to Every POD Group:

```yaml
pod_number: 1  # leaf-spine POD number, starts with 1

spine:
  # list of spine interfaces used as uplinks to super-spines
  # taking `max_spine_to_super_spine_links` into account
  # for example: spine1, spine2, spine3, ...
  # or spine1, spine1, spine2, spine2, etc.
  uplinks_to_super_spine_interfaces: ['Ethernet10', 'Ethernet11', 'Ethernet12', 'Ethernet13']
  nodes:
    <spine-hostname>:
      # super-spine interfaces to spines
      # taking `max_spine_to_super_spine_links` into account
      # for example: super-spine1, super-spine2, super-spine3, ...
      # or super-spine1, super-spine1, super-spine2, super-spine2, etc.
      super_spine_interfaces: ['Ethernet1', 'Ethernet1', 'Ethernet1', 'Ethernet1']
    <-- etc. -->

# Point to Point Network Summary range, assigned as /31 for each
# uplink interfaces
# Assign range larger then total
# [ max_spines_in_a_POD * max_super_spines * max_spine_to_super_spine_links * 2 ]
super_spine_underlay_p2p_network_summary: 172.31.1.0/24
```

Following variables must be now defined on DC and not POD level:

- `p2p_uplinks_mtu`
- `bgp_peer_groups`

## Role Enchancements for dedicated Overlay Controllers

This enhancement will allow support for dedicated Overlay Controllers connected to fabric nodes.
Overlay Controllers can be connected to any other device type.
Overlay Controllers can currently only be used as EVPN Overlay Controllers

### Inventory Structure

The inventory must have a dedicated group for Overlay Controllers. Example:

```yaml
all:
  children:
    < DC-group-name >:
      children:
        < Overlay Controllers group name >:
          hosts:
            < Overlay Controller name >:
              ansible_host: < management IP >
            < Overlay Controller name >:
              ansible_host: < management IP >
            ...
```

### Additional Variables Required For Overlay Controllers Deployment

Defaults:
```yaml
max_overlay_controller_to_switch_links: 1
```

Assigned to the DC group:
 
```yaml
overlay_controller:
  platform: <platform>   # overlay-controller platform
  defaults: #Default variables, can be overridden when defined under each node
    remote_switches: [ <switch_inventory_hostname> , <switch_inventory_hostname> ] #Remote Switches connected to uplink interfaces
    uplink_to_remote_switches: [ <uplink_interface> , <uplink_interface> ]
    bgp_as: <BGP AS>
  nodes:
    <inventory_hostname>:
      id: <number> # Starting from 1
      mgmt_ip: < IPv4_address/Mask >
      remote_switches_interfaces: [ <remote_switch_interface> , <remote_switch_interface> ] # Interfaces on remote switch

# Point to Point Network Summary range, assigned as /31 for each uplink interfaces
# Assign range larger than [ total overlay_controllers * max_overlay_controller_to_switch_links * 2]
overlay_controller_p2p_network_summary: < IPv4_network/Mask >

# IP address summary for BGP evpn overlay peering loopback for Overlay Controllers | Required
# Assigned as /32 to Loopback0
# Assign range larger then:
# [ total overlay_controllers ]
overlay_controller_loopback_network_summary: < IPv4_network/Mask >

# additional lines for overlay-controller BGP config
overlay_controller_bgp_defaults:
  #  - update wait-for-convergence
  #  - update wait-install
  - no bgp default ipv4-unicast
  - distance bgp 20 200 200
  - graceful-restart restart-time 300
  - graceful-restart
```

Assigned to Overlay Controller Group:

```yaml
type: overlay-controller # identifies every host in the group as overlay-controller
```

## Role Enhancements for Flexible EVPN Overlay peering design

This enhancement will allow a more flexible EVPN Overlay peering design.
Overlay peerings can be configured to any of the following options as well as combinations thereof:
* l3leaf <-> spine
* l3leaf <-> super-spine
* l3leaf <-> overlay-controller
* spine <-> spine (in other PODs)
* spine <-> super-spine
* spine <-> overlay-controller
* super-spine <-> super-spine (in other DCs)
* super-spine <-> overlay-controller
* overlay-controller <-> overlay-controller (in other DCs)
The overlay peerings will be derived from the variable `evpn_overlay_controller_groups` on the "client".
Ex. setting `evpn_overlay_controller_groups : [ DC1_SUPER_SPINES ]` in the `DC1-POD1-L3LEAFS.yml` group_var file will setup evpn peerings between 
all the l3leafs in this group and all devices in the group `DC1_SUPER_SPINES`. The devices in group `DC1_SUPER_SPINES` will detect that others are
pointing at them, and configure the peering as well.

Currently all spines, super-spines and overlay-controllers will have the EVPN adress-family and EVPN BGP peer-group configured even if they have no
EVPN overlay peerings.

To be backward compatible the templates will use the old behavior of configuring EVPN Overlay between all spines and l3leafs if `evpn_overlay_controller_groups`
is not defined on the l3leafs group.

### Inventory Structure

There are no specific requirements for this feature, so this is just an example for reference.

```yaml
all:
  children:
    < DC-group-name >:
      children:
        < Overlay Controllers group name >:
          <-- omitted -->
        < Super Spines group name >:
          <-- omitted -->
        < DC POD 1 group name >:
          children:
            < spines group >:
              <-- omitted -->
            < leaf group >:
              <-- omitted -->
```

### Additional Variables For Flexible EVPN Overlay peerings

Assigned to any device group:

```yaml
evpn_overlay_controller_groups: [ < inventory_group > ]  # One or more groups containing EVPN RR/RS.
```

## How to test

https://github.com/ClausHolbechArista/claus-l3ls-dev/tree/superspine-alternative

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-l3leaf-yml.j2 [hunk 1]
```diff
@@ -159,6 +159,35 @@
 {%         endif %}
 {%     endif %}
 {% endif %}
+{# Gather overlay peering information #}
+{% set switch.evpn_overlay_peers = [] %}
+{# #}
+{# Look for evpn_overlay_controllers / servers where we are client #}
+{% if evpn_overlay_controller_groups is defined and evpn_overlay_controller_groups is not none %}
+{%     for evpn_overlay_controller_group in evpn_overlay_controller_groups %}
+{%         for some_switch in groups[evpn_overlay_controller_group] %}
```
Comment (titom73) [inline ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-l3leaf-yml.j2]: Maybe we should use a more specific name to relate `some_switch` ? Like `evpn_overlay_switch` as it is only iteration over `groups[evpn_overlay_controller_group]`

File: ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-overlay-controller-yml.j2 [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+{%                 endif %}
+{%                 do switch.evpn_overlay_peers.append(peer) %}
+{%             endif %}
+{%         endfor %}
+{%     endif %}
+{% endfor %}
+{# Look for evpn_overlay_controllers / servers where we are client #}
+{% if evpn_overlay_controller_groups is defined and evpn_overlay_controller_groups is not none %}
+{%     for evpn_overlay_controller_group in evpn_overlay_controller_groups %}
+{%         for some_switch in groups[evpn_overlay_controller_group] %}
```
Comment (titom73) [inline ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-overlay-controller-yml.j2]: Maybe we should use a more specific name to relate `some_switch` ? Like `evpn_overlay_switch` as it is only iteration over `groups[evpn_overlay_controller_group]`

File: ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-spine-yml.j2 [hunk 3]
```diff
@@ -8,6 +8,78 @@
...
+{%                 endif %}
+{%                 do switch.evpn_overlay_peers.append(peer) %}
+{%             endif %}
+{%         endfor %}
+{%     endif %}
+{% endfor %}
+{# Look for evpn_overlay_controllers / servers where we are client #}
+{% if evpn_overlay_controller_groups is defined and evpn_overlay_controller_groups is not none %}
+{%     for evpn_overlay_controller_group in evpn_overlay_controller_groups %}
+{%         for some_switch in groups[evpn_overlay_controller_group] %}
```
Comment (titom73) [inline ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-spine-yml.j2]: Maybe we should use a more specific name to relate `some_switch` ? Like `evpn_overlay_switch` as it is only iteration over `groups[evpn_overlay_controller_group]`

File: ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-super-spine-yml.j2 [hunk 4]
```diff
@@ -8,6 +8,78 @@
...
+{%                 endif %}
+{%                 do switch.evpn_overlay_peers.append(peer) %}
+{%             endif %}
+{%         endfor %}
+{%     endif %}
+{% endfor %}
+{# Look for evpn_overlay_controllers / servers where we are client #}
+{% if evpn_overlay_controller_groups is defined and evpn_overlay_controller_groups is not none %}
+{%     for evpn_overlay_controller_group in evpn_overlay_controller_groups %}
+{%         for some_switch in groups[evpn_overlay_controller_group] %}
```
Comment (titom73) [inline ansible_collections/arista/avd/roles/eos_l3ls_evpn/templates/evpn-fabric-super-spine-yml.j2]: Maybe we should use a more specific name to relate `some_switch` ? Like `evpn_overlay_switch` as it is only iteration over `groups[evpn_overlay_controller_group]`

**Review by titom73 — APPROVED** [review]
Looks very promising

**Comment by ClausHolbechArista** [comment]
Too many conflicts. Moved to new branch.

**Pull request closed without merging**